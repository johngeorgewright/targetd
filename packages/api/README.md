# @targetd/api

Load and query data

## Installation

| JS Runtime | Command                     |
| ---------- | --------------------------- |
| Node.js    | `npx jsr add @targetd/api`  |
| Bun        | `bunx jsr add @targetd/api` |
| Deno       | `deno add jsr:@targetd/api` |
