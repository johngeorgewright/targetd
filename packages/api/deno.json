{
  "name": "@targetd/api",
  "version": "7.1.0",
  "exports": "./src/index.ts",
  "fmt": {
    "semiColons": false,
    "singleQuote": true
  },
  "tasks": {
    "pwd": "echo $(pwd)",
    "test": "deno test --allow-all --trace-leaks"
  },
  "imports": {
    "@myty/promise-chain": "jsr:@myty/promise-chain@^0.4.6",
    "fast-deep-equal": "npm:fast-deep-equal@^3.0.0",
    "ts-toolbelt": "npm:ts-toolbelt@^9.0.0",
    "zod": "npm:zod@^4.0.15"
  },
  "publish": {
    "include": [
      "deno.json",
      "src/",
      "LICENSE",
      "README.md"
    ]
  }
}
