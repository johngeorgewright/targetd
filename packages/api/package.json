{
  "name": "@targetd/api",
  "version": "6.0.1",
  "description": "Load and query data",
  "main": "dist/index.js",
  "sideEffects": false,
  "scripts": {
    "build": "yarn clean && yarn tsc",
    "clean": "yarn rimraf dist",
    "start": "yarn tsc --watch --preserveWatchOutput",
    "test": "yarn jest --passWithNoTests"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/config/issues"
  },
  "homepage": "https://github.com/johngeorgewright/config#readme",
  "devDependencies": {
    "@semantic-release/commit-analyzer": "11.1.0",
    "@semantic-release/exec": "6.0.3",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "9.2.6",
    "@semantic-release/release-notes-generator": "12.1.0",
    "@types/jest": "29.5.13",
    "@types/node": "20.16.9",
    "jest": "29.7.0",
    "prettier-2": "npm:prettier@^2",
    "rimraf": "6.0.1",
    "semantic-release": "23.1.1",
    "semantic-release-monorepo": "8.0.2",
    "ts-jest": "29.2.5",
    "typescript": "5.4.5",
    "zod": "3.23.8"
  },
  "dependencies": {
    "fast-deep-equal": "3.1.3",
    "ts-toolbelt": "9.6.0",
    "tslib": "2.7.0"
  },
  "peerDependencies": {
    "zod": "^3.20.0"
  }
}
