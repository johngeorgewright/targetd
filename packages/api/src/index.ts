export { default as Data } from './Data.ts'
export { DataItemParser } from './parsers/DataItem.ts'
export { DataItemRuleParser } from './parsers/DataItemRule.ts'
export { DataItemsParser } from './parsers/DataItems.ts'
export { default as createTargetingDescriptor } from './createTargetingDescriptor.ts'
export type { default as TargetingDescriptor } from './parsers/TargetingDescriptor.ts'
export type { default as TargetingPredicate } from './parsers/TargetingPredicate.ts'
export type { default as TargetingPredicates } from './parsers/TargetingPredicates.ts'
export { equalsPredicate, targetEquals } from './predicates/equals.ts'
export {
  targetIncludes,
  targetIncludesPredicate,
} from './predicates/targetIncludes.ts'
export type { StaticRecord } from './types.ts'
export type * as DT from './types/Data.ts'
export type * as FTTT from './types/FallThroughTargeting.ts'
export type * as PT from './types/Payload.ts'
export type * as QT from './types/Query.ts'
export type * as TT from './types/Targeting.ts'
