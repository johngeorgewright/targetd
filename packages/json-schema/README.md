# @targetd/json-schema

> Creates JSON schema from targetd data objects

## Installation

| JS Runtime | Command                                          |
| ---------- | ------------------------------------------------ |
| Node.js    | `npx jsr add @targetd/api @targetd/json-schema`  |
| Bun        | `bunx jsr add @targetd/api @targetd/json-schema` |
| Deno       | `deno add @targetd/api @targetd/json-schema`     |
