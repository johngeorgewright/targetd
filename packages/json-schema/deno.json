{
  "name": "@targetd/json-schema",
  "version": "3.0.1",
  "exports": "./src/index.ts",
  "imports": {
    "yargs": "npm:yargs@^17.7.2",
    "zod": "npm:zod@^3.20.0",
    "zod-to-json-schema": "npm:zod-to-json-schema@^3.23.3"
  },
  "fmt": {
    "semiColons": false,
    "singleQuote": true
  },
  "tasks": {
    "test": "deno test --allow-all --trace-leaks"
  }
}
