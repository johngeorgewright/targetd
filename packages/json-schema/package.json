{
  "name": "@targetd/json-schema",
  "version": "3.0.1",
  "description": "Creates JSON schema from targetd data objects",
  "main": "dist/index.js",
  "sideEffects": false,
  "bin": {
    "targetd-json-schema": "./dist/cli.js"
  },
  "scripts": {
    "build": "yarn clean && yarn tsc",
    "clean": "yarn rimraf dist",
    "start": "yarn tsc --watch --preserveWatchOutput",
    "release": "yarn semantic-release -e semantic-release-monorepo",
    "test": "yarn jest --passWithNoTests"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/config/issues"
  },
  "homepage": "https://github.com/johngeorgewright/config#readme",
  "devDependencies": {
    "@semantic-release/commit-analyzer": "11.1.0",
    "@semantic-release/exec": "6.0.3",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "9.2.6",
    "@semantic-release/release-notes-generator": "12.1.0",
    "@targetd/api": "workspace:*",
    "@types/jest": "29.5.13",
    "@types/node": "20.16.9",
    "@types/yargs": "^17.0.32",
    "jest": "29.7.0",
    "rimraf": "6.0.1",
    "semantic-release": "23.1.1",
    "semantic-release-monorepo": "8.0.2",
    "ts-jest": "29.2.5",
    "ts-node": "10.9.2",
    "typescript": "5.4.5",
    "zod": "3.23.8"
  },
  "dependencies": {
    "tslib": "2.7.0",
    "yargs": "17.7.2",
    "zod-to-json-schema": "3.23.3"
  },
  "peerDependencies": {
    "@targetd/api": ">=6.0.1",
    "ts-node": "^10.9.2",
    "zod": "^3.20.0"
  }
}
