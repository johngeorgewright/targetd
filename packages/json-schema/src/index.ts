import { DataItemParser, DataItemsParser, type DT } from '@targetd/api'
import { string } from 'zod'
import $zodToJSONSchema from 'zod-to-json-schema'

// Incorrect typing
const zodToJSONSchema =
  $zodToJSONSchema as unknown as typeof $zodToJSONSchema.default

export function dataJSONSchemas<D extends DT.Any>(data: D) {
  return zodToJSONSchema(
    DataItemsParser(
      data.payloadParsers,
      data.targetingParsers,
      data.fallThroughTargetingParsers,
    ).extend({ $schema: string().optional() }),
    {
      effectStrategy: 'input',
    },
  )
}

export function dataJSONSchema<D extends DT.Any>(
  data: D,
  name: keyof DT.PayloadParsers<D>,
) {
  return zodToJSONSchema(
    DataItemParser(
      data.payloadParsers[name],
      data.targetingParsers,
      data.fallThroughTargetingParsers,
    ).extend({ $schema: string().optional() }),
    {
      effectStrategy: 'input',
    },
  )
}
