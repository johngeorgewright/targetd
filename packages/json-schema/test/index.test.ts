import { assertSnapshot } from 'jsr:@std/testing/snapshot'
import { test } from 'jsr:@std/testing/bdd'
import { Data, targetIncludes } from '@targetd/api'
import { dataJSONSchemas } from '@targetd/json-schema'
import { z } from 'zod/v4'

test('json schema for simple data object', async (t) => {
  await assertSnapshot(
    t,
    dataJSONSchemas(
      await Data.create({
        data: {
          foo: z.string(),
        },
        targeting: {
          weather: targetIncludes(z.string()),
        },
        fallThroughTargeting: {
          browser: targetIncludes(z.string()),
        },
      }).addRules('foo', [
        {
          payload: 'bar',
        },
      ]),
    ),
  )
})
