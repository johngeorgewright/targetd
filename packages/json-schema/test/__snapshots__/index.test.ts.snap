export const snapshot = {};

snapshot[`json schema for simple data object 1`] = `
{
  "\$schema": "http://json-schema.org/draft-07/schema#",
  additionalProperties: false,
  properties: {
    "\$schema": {
      type: "string",
    },
    foo: {
      additionalProperties: false,
      properties: {
        rules: {
          items: {
            additionalProperties: false,
            properties: {
              payload: {
                type: "string",
              },
              targeting: {
                anyOf: [
                  {
                    additionalProperties: false,
                    properties: {
                      browser: {
                        items: {
                          type: "string",
                        },
                        type: "array",
                      },
                      weather: {
                        items: {
                          type: "string",
                        },
                        type: "array",
                      },
                    },
                    type: "object",
                  },
                  {
                    items: {
                      "\$ref": "#/properties/foo/properties/rules/items/properties/targeting/anyOf/0",
                    },
                    type: "array",
                  },
                ],
              },
            },
            required: [
              "payload",
            ],
            type: "object",
          },
          type: "array",
        },
      },
      required: [
        "rules",
      ],
      type: "object",
    },
  },
  type: "object",
}
`;
